In [1]:
import os
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
caminho = os.path.join('..', 'dados', 'Data Sets- STATA')

## C1

In [2]:
filename = os.path.join(caminho, 'VOTE1.DTA')
dfc1 = pd.read_stata(filename, preserve_dtypes=False)
dfc1.head()

state  district  democA  voteA     expendA     expendB  prtystrA  lexpendA  \
0    AL         7       1     68  328.295990    8.737000        41  5.793916   
1    AK         1       0     62  626.377014  402.476990        60  6.439952   
2    AZ         2       1     73   99.607002    3.065000        55  4.601233   
3    AZ         3       0     69  319.690002   26.281000        64  5.767352   
4    AR         3       0     75  159.220993   60.054001        66  5.070293   

   lexpendB     shareA  
0  2.167567  97.407669  
1  5.997638  60.881039  
2  1.120048  97.014763  
3  3.268846  92.403702  
4  4.095244  72.612473

## i
b1 é a força do parâmetro expendA. 1% de expansão em expendA aumenta b1 voteA
## ii
H0: b1 == -b2

In [54]:
modelc1 = smf.ols(formula='voteA ~ np.log(expendA) + np.log(expendB) + prtystrA', data=dfc1)
regrc1 = modelc1.fit()

In [55]:
print(f"voteA = {regrc1.params['Intercept']:.3f} + " +
      f"{regrc1.params['np.log(expendA)']:.3f}log(expendA) + " + 
      f"{regrc1.params['np.log(expendB)']:.3f}expendB + " +
      f"{regrc1.params['prtystrA']:.3f}prtystrA")
print(f"\t({regrc1.bse['Intercept']:.3f}) \t({regrc1.bse['np.log(expendA)']:.3f}) " + 
      f"\t({regrc1.bse['np.log(expendB)']:.3f}) \t({regrc1.bse['prtystrA']:.3f})")

print(f'n = {len(dfc1)}, R^2 = {regrc1.rsquared:.4f}')

voteA = 45.079 + 6.083log(expendA) + -6.615expendB + 0.152prtystrA
	(3.926) 	(0.382) 	(0.379) 	(0.062)
n = 173, R^2 = 0.7926


## iv


In [61]:
modelc1b = smf.ols(formula='voteA ~ np.log(expendA) + I(np.log(expendB) - np.log(expendA)) + prtystrA', data=dfc1)
regrc1b = modelc1b.fit()

In [63]:
print(f"voteA = {regrc1b.params['Intercept']:.3f} + " +
      f"{regrc1b.params['np.log(expendA)']:.3f}O + "  +
      f"{regrc1b.params['I(np.log(expendB) - np.log(expendA))']:.3f}log(expendB)-log(expendA) + " +
      f"{regrc1b.params['prtystrA']:.3f}prtystrA")
print(f"\t({regrc1b.bse['Intercept']:.3f}) \t({regrc1b.bse['np.log(expendA)']:.3f}) " + 
      f"\t({regrc1b.bse['I(np.log(expendB) - np.log(expendA))']:.3f}) \t({regrc1b.bse['prtystrA']:.3f})")
print(f'n = {len(dfc1)}, R^2 = {regrc1b.rsquared:.4f}')

voteA = 45.079 + -0.532O + -6.615log(expendB)-log(expendA) + 0.152prtystrA
	(3.926) 	(0.533) 	(0.379) 	(0.062)
n = 173, R^2 = 0.7926


h0: b1 == -b2, O = 0, O = b1 - b2 (Para estimar erro padrão de h0, estimar erro padrão de b1 - b2 = 0)

When we estimate this equation we obtain θ  1 ≈ –.532 and se( θ  1 ) ≈ .533.
The t statistic for the hypothesis in part (ii) is –.532/.533 ≈ –1. Therefore, we fail to reject H 0 : β 2 = – β 1 .

In [60]:
print(regrc1.summary())
print(regrc1b.summary())

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Wed, 11 Jul 2018   Prob (F-statistic):           1.76e-57
Time:                        13:06:41   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          45.0789      3.926     

In [ ]:
t